runtime/j9vm/jvm.c

```diff
@@ -4549,27 +4590,30 @@ JVM_RegisterSignal(jint sigNum, void* handler)
 		/* Don't allow user to register a native handler since
 		 * the signal is already used by the VM.
 		 */
-	} else if (J9_ARE_NO_BITS_SET(javaVM->sigFlags, J9_SIG_XRS)
-			&& isSignalSpecial(sigNum)) {
+		return (void *)J9_SIG_ERR;
+	} else if (isSpecialSignal && J9_ARE_ANY_BITS_SET(javaVM->sigFlags, J9_SIG_XRS | J9_SIG_XRS_ASYNC)) {
 		/* Don't allow user to register a native handler since
 		 * the signal is already used by the VM.
 		 */
+		return (void *)J9_SIG_ERR;
+	} else if (isSpecialSignal && isSignalIgnored(sigNum)) {
+		return (void *)J9_SIG_IGNORED;
 	} else {
 		/* Register the signal */
 #if defined(WIN32)
 		if ((void *)J9_PRE_DEFINED_HANDLER_CHECK == handler) {
-			handler = (void *)dummySignalHandler;
+			handler = (void *)defaultSignalHandler;
 		}
 		oldHandler = OMRSIG_SIGNAL(sigNum, handler);
 #else /* defined(WIN32) */
-		sigemptyset(&newSignalAction.sa_mask);
+		sigfillset(&newSignalAction.sa_mask);
 #if !defined(J9ZTPF)
-		newSignalAction.sa_flags = SA_RESTART;
+		newSignalAction.sa_flags = SA_RESTART | SA_SIGINFO;
```

Comment:
Not sure why SA_SIGINFO is being set. This means to use sa_sigaction (which takes 2 args) which this code is not doing.

Reply:
I was trying to determine the functionality of `defaultSignalHandler`. So, I tried existing handlers such as `omrsig_handler` and `structuredSignalHandlerVM`, which had three input arguments. `JVM_RegisterSignal` would register a `handler` for a specific signal. When `JVM_RaiseSignal` is called for a signal, then the `defaultSignalHandler` is invoked. I found that the job of `defaultSignalHandler` is to invoke the registered handler for the signal. `defaultSignalHandler` invokes the registered handler for the signal by calling `Signal.dispatch`. I will remove `SA_SIGINFO` since `defaultSignalHandler` only takes one input argument. but, things still work with `SA_SIGINFO`.